### ISSUE TITLE
Candidate g gives negative 'Betti numbers' (part 1)

### ISSUE THREAD

**Issue by jonathanfine** [jonathanfine/py-linhomy#16]
See below.  There's one line that sticks out because of a minus sign.

Unless there's an error in the implementation (eg consequence of integer overflow) this means that the algorithm is incorrect.

```python
py-linhomy/py$ python3 -i -m linhomy.matrices
>>> for v in sorted(set(_cache.values())):
...     if len(v) == 34:
...         print(''.join(map('{0:x}'.format, numpy.dot(g_from_F[8], v))))
... 
1000000000000000000000000000000000
1111010010000100000001000000000000
1011010010000100000001000000000000
1242032132010320001002200000010000
1001010010000100000001000000000000
1124020031210312001002120000010000
1000010010000100000001000000000000
1112040020000311201002112000010000
1022040020000302201002022000010000
1000000010000100000001000000000000
1113130130010300001002000000010000
1111020040000200000002111020010000
1012020040000200000002012020010000
1000000000000100000001000000000000
1121031131010310001002100000010000
1244182352020742314113323100031010
1000000000000000000001000000000000
1112010031110311001002110000010000
1252053296040533002003440043232-110
1126140181230512002003134130130110
1111020010000311101002111000010000
1011030130010300001002000000010000
1111010020000100000002111010010000
1012010030110301001002010000010000
1124040181230512002003232030131110
1011020010000301101002011000010000
1011010020000100000002011010010000
1001010030010300001002000000010000
1001020010000300101002001000010000
1001010020000100000002001010010000
1000010010000300001002000000010000
1000010020000100000002000010010000
1114130171141711004013110000030010
1113160230010711314013113100030010
1000000010000100000002000000010000
1113130160020300001003113130130010
1121041131010732104113221000031010
1121031162020311001003221031131010
1112010041110311001003222010031110
1011040130010701104013011000030010
1011030160020300001003011030130010
1012010040110301001003022010030110
1001010040010300001003001010030010
11141301a1151711004014224130161141
>>> 

```

**Comment by jonathanfine** [comment]
Here we zoom in on the error.
```python
>>> for k, v in sorted(set(_cache.items())):
...     if len(v) == 34:
...         print(k, ''.join(map('{0:x}'.format, numpy.dot(g_from_F[8], v))))
... 
[...]
b'J(IC,ICCIC)' 1252053296040533002003440043232-110
[...]
```

The error affects g_i, for i = 20:01 - see below
```python
>>> FIBWORDS[8][-3]
b'\x02\x02\x01\x02\x01'
```

**Comment by jonathanfine** [comment]
The negative value is the result of the dot product of a CD-vector with a row in g_from_CD[8].  Here's how:
```python
>>> _cache[b'J(IC,ICCIC)']
(1, 16, 84, 216, 664, 321, 1388, 2120, 296, 1724, 3616, 3546, 10960, 172, 1316, 3588, 
4773, 14756, 3357, 14588, 22280, 61, 604, 2076, 3539, 10932, 3357, 14588, 22280, 
1808, 10568, 22168, 21720, 67200)
>>> 
>>>  # Compute the CD-vector.
>>> ' '.join(map(str, numpy.dot(CD_from_F[8], _cache[b'J(IC,ICCIC)'])))
'1 2 0 0 0 1 3 -1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 3 -1 -1 -1 0 0 0'
>>> 
>>> # Compute the relevant row in g_from_CD.
>>> ' '.join(map(str, g_from_CD[8][-3,:]))
'0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0'
```

To fix this change g_from_CD[8][-3,:]. The only sensible change is to have a positive entry in the CD-vector contribute to the g_i. This means changing a zero to one in the row in g_from_CD[8].  The possible places for the change are:
```python
>>> for i, c in enumerate(numpy.dot(CD_from_F[8], _cache[b'J(IC,ICCIC)'])):
...     if c > 0:
...         print(i, c, FIBWORDS[8][i])
... 
0 1 b'\x01\x01\x01\x01\x01\x01\x01\x01'
1 2 b'\x01\x01\x01\x01\x01\x01\x02'
5 1 b'\x01\x01\x01\x02\x01\x01\x01'
6 3 b'\x01\x01\x01\x02\x01\x02'
22 1 b'\x02\x01\x01\x01\x01\x02'
26 1 b'\x02\x01\x02\x01\x01\x01'
27 3 b'\x02\x01\x02\x01\x02'
```

**Comment by jonathanfine** [comment]
We want an extra contribution to **20:01**.  I think **10:00:00** is the most likely candidate to provide this.  It is closest to the target (in some sense) and is already an example of the hard D-rule.  In fact, it is the first such example.  Perhaps we've got hard-D wrong. (We never had any real evidence for it, other than one example in n=5.)

We could modify hard-D to replace the current **10:00:00** output of  **00:01:01** by the desired **20:01**.  Doing this depends on #17.

**Comment by jonathanfine** [comment]
Mostly done - 7 negative 'Betti numbers' in dimension 10 left to do.  Merge into trunk and closed, with residue places in new issue #20.
